### PR TITLE
Ignore Initial and Handshake packet from a different IP address and port

### LINKED ISSUE #1116: Crash in [@ neqo_transport::path::Path::set_remote_cid]
More info can be found in bug [1696706](https://bugzilla.mozilla.org/show_bug.cgi?id=1696706)

### PR DESCRIPTION
**Pull request by ddragana**
Fixes #1116

### CODE REVIEW

**Comment by ddragana** [comment]
The crash can be triggered when IP address of a Initial packet is different from the address a client have sent its Initial packet to.

neqo is happy to accept Handshake packet from any address, butt I decided to restrict that as well.

File: neqo-transport/src/connection/mod.rs [hunk 1]
```diff
@@ -1090,6 +1091,17 @@ impl Connection {
             return Ok(PreprocessResult::Next);
         }
 
+        if (packet.packet_type() == PacketType::Initial
+            || packet.packet_type() == PacketType::Handshake)
+            && self.role == Role::Client
+            && !path.borrow().is_primary()
+        {
+            // If we have received a packet from a different address than we have sent to
+            // we should ignore the packet. In such a case a path will be newly created
```
Comment (martinthomson) [inline neqo-transport/src/connection/mod.rs:1100]:
```suggestion
            // we should ignore the packet. In such a case a path will be a newly created
```

File: neqo-transport/src/connection/tests/handshake.rs [hunk 2]
```diff
@@ -828,3 +829,55 @@ fn garbage_initial() {
...
+    assert!(out.as_dgram_ref().is_some());
+    assert_eq!(out.as_dgram_ref().unwrap().len(), PATH_MTU_V6);
+
+    let mut server = default_server();
+    let out = server.process(out.dgram(), now());
+    assert!(out.as_dgram_ref().is_some());
+    assert_eq!(out.as_dgram_ref().unwrap().len(), PATH_MTU_V6);
+
+    let (s_in, s_hs) = split_datagram(&out.dgram().unwrap());
+    assert!(s_hs.is_some());
```
Comment (martinthomson) [inline neqo-transport/src/connection/tests/handshake.rs:869]:
```suggestion
```
You unwrap a few lines down, so this is OK to skip.

File: neqo-transport/src/connection/tests/handshake.rs [hunk 3]
```diff
@@ -828,3 +829,55 @@ fn garbage_initial() {
...
+fn drop_handshake_packet_from_wrong_address() {
+    let mut client = default_client();
+    let out = client.process(None, now());
+    assert!(out.as_dgram_ref().is_some());
+    assert_eq!(out.as_dgram_ref().unwrap().len(), PATH_MTU_V6);
+
+    let mut server = default_server();
+    let out = server.process(out.dgram(), now());
+    assert!(out.as_dgram_ref().is_some());
+    assert_eq!(out.as_dgram_ref().unwrap().len(), PATH_MTU_V6);
```
Comment (martinthomson) [inline neqo-transport/src/connection/tests/handshake.rs:866]:
```suggestion
```

File: neqo-transport/src/connection/tests/handshake.rs [hunk 4]
```diff
@@ -828,3 +829,55 @@ fn garbage_initial() {
...
+    let out = client.process(Some(dgram), now());
+    assert!(out.as_dgram_ref().is_none());
+}
+
+#[test]
+fn drop_handshake_packet_from_wrong_address() {
+    let mut client = default_client();
+    let out = client.process(None, now());
+    assert!(out.as_dgram_ref().is_some());
+    assert_eq!(out.as_dgram_ref().unwrap().len(), PATH_MTU_V6);
```
Comment (martinthomson) [inline neqo-transport/src/connection/tests/handshake.rs:861]:
```suggestion
```

File: neqo-transport/src/connection/tests/handshake.rs [hunk 5]
```diff
@@ -828,3 +829,55 @@ fn garbage_initial() {
...
+fn drop_initial_packet_from_wrong_address() {
+    let mut client = default_client();
+    let out = client.process(None, now());
+    assert!(out.as_dgram_ref().is_some());
+    assert_eq!(out.as_dgram_ref().unwrap().len(), PATH_MTU_V6);
+
+    let mut server = default_server();
+    let out = server.process(out.dgram(), now());
+    assert!(out.as_dgram_ref().is_some());
+    assert_eq!(out.as_dgram_ref().unwrap().len(), PATH_MTU_V6);
```
Comment (martinthomson) [inline neqo-transport/src/connection/tests/handshake.rs:843]:
```suggestion
```

File: neqo-transport/src/connection/tests/handshake.rs [hunk 6]
```diff
@@ -828,3 +829,55 @@ fn garbage_initial() {
     let garbage = Datagram::new(addr(), addr(), corrupted);
     assert_eq!(Output::None, server.process(Some(garbage), now()));
 }
+
+#[test]
+fn drop_initial_packet_from_wrong_address() {
+    let mut client = default_client();
+    let out = client.process(None, now());
+    assert!(out.as_dgram_ref().is_some());
+    assert_eq!(out.as_dgram_ref().unwrap().len(), PATH_MTU_V6);
```
Comment (martinthomson) [inline neqo-transport/src/connection/tests/handshake.rs:838]: No need to check lengths here.
```suggestion
```

File: neqo-transport/src/connection/mod.rs [hunk 7]
```diff
@@ -1090,6 +1091,17 @@ impl Connection {
...
+            // temporary path, not a primary path.
```
Comment (martinthomson) [inline neqo-transport/src/connection/mod.rs:1101]:
```suggestion
            // temporary path, not the primary path.
```

**Review by martinthomson — COMMENTED** [review]
A few nits, but this is good.

**Pull request merged**